deepchem/data/data_loader.py

```diff
@@ -437,6 +439,192 @@ def _featurize_shard(self, shard):
+      input_files = [input_files]
+
+    def shard_generator():
+      """Yield X, y, w, and ids for shards."""
+      for shard_num, shard in enumerate(
+          self._get_shards(input_files, shard_size)):
+
+        time1 = time.time()
+        X, valid_inds = self._featurize_shard(shard)
+        if self.id_field:
+          ids = shard[self.id_field].values
+        else:
+          ids = np.ones(len(X))
+        ids = ids[valid_inds]
+
+        if len(self.tasks) > 0:
+          # Featurize task results iff they exist.
+          y, w = _convert_df_to_numpy(shard, self.tasks)
+
+          if self.label_field:
+            y = shard[self.label_field]
+          if self.weight_field:
+            w = shard[self.weight_field]
+
+          # Filter out examples where featurization failed.
+          y, w = (y[valid_inds], w[valid_inds])
+          assert len(X) == len(ids) == len(y) == len(w)
+        else:
+          # For prospective data where results are unknown, it
+          # makes no sense to have y values or weights.
+          y, w = (None, None)
+          assert len(X) == len(ids)
+
+        time2 = time.time()
+        logger.info("TIMING: featurizing shard %d took %0.3f s" %
+                    (shard_num, time2 - time1))
+        yield X, y, w, ids
+
+    return DiskDataset.create_dataset(shard_generator(), data_dir)
+
+  def _get_shards(self, input_files, shard_size):
+    """Defines a generator which returns data for each shard"""
+    return load_json_files(input_files, shard_size)
+
+  def _featurize_shard(self, shard):
+    """Featurizes a shard of an input dataframe."""
+    return self._featurize_df(
+        shard, self.featurizer, log_every_n=self.log_every_n)
+
+  def _featurize_df(self,
+                    shard,
+                    featurizer: Featurizer,
+                    log_every_n: int = 1000) -> Tuple[np.ndarray, np.ndarray]:
+    """Featurize individual samples in dataframe.
+
+    Helper that given a featurizer that operates on individual
+    samples, computes & adds features for that sample to the 
+    features dataframe.
+
+    Parameters
+    ----------
+    shard: pd.DataFrame
+      DataFrame that holds data to be featurized.
+    featurizer: Featurizer
+      An instance of `dc.feat.Featurizer`.
+    log_every_n: int, optional (default 1000)
+      Emit a logging statement every `log_every_n` rows.
+
+    Returns
+    -------
+    features : np.ndarray
+      Array of feature vectors.
+    valid_inds : np.ndarray
+      Boolean values indicating successfull featurization.
+
+    """
+
+    features = []
+    field = self.feature_field
+    data = shard[field].tolist()
+    for idx, datapoint in enumerate(data):
+      features.append(featurizer.featurize([datapoint]))
+
+    valid_inds = np.array(
+        [1 if elt.size > 0 else 0 for elt in features], dtype=bool)
+    features = [
+        elt for (is_valid, elt) in zip(valid_inds, features) if is_valid
+    ]
```

Comment:
it is possible to merge three loops to one loop.

```python
features = []
valid_inds = []
for idx, datapoint in enumerate(data):
    feat = featurizer.featurize([datapoint])
    is_valid = 1 if feat.size > 0 else 0
    valid_inds.append(is_valid)
    if is_valid == 1:
        features.append(feat)
```

Reply:
Done!